lib/ansible/executor/process/forking.py

```diff
@@ -37,17 +39,103 @@
     pass
 
 from ansible.errors import AnsibleConnectionFailure
+from ansible.executor.process.model import ProcessModelBase, ResultsSentinel
 from ansible.executor.task_executor import TaskExecutor
 from ansible.executor.task_result import TaskResult
 from ansible.module_utils._text import to_text
+from ansible.module_utils.six.moves import queue as Queue
 
 try:
     from __main__ import display
 except ImportError:
     from ansible.utils.display import Display
     display = Display()
 
-__all__ = ['WorkerProcess']
+__all__ = ['ProcessModelForking', 'WorkerProcess']
+
+
+def results_thread_main(pm):
+    while True:
+        try:
+            result = pm._worker_results_q.get()
+            if isinstance(result, ResultsSentinel):
+                break
+            else:
+                pm._final_q_lock.acquire()
+                pm._final_q.append(result)
+                pm._final_q_lock.release()
+        except (IOError, EOFError) as e:
+            break
+        except Queue.Empty:
+            pass
+
+
+class ProcessModelForking(ProcessModelBase):
+    def __init__(self, tqm):
+        super(ProcessModelForking, self).__init__(tqm)
+
+        # save a couple of things from the TQM we need for forking
+        self._loader = tqm._loader
+        self._variable_manager = tqm._variable_manager
+
+        # create the worker results queue, which is where all workers
+        # send their results. The results thread reads from this in the
+        # background and copies the results to the final_q deque object
+        self._worker_results_q = multiprocessing.Queue()
+
+        # create the result processing thread for reading results in the background
+        self._results_thread = threading.Thread(target=results_thread_main, args=(self,))
+        self._results_thread.daemon = True
+        self._results_thread.start()
+
+    def initialize_workers(self, num):
+        self._cur_worker = 0
+        self._workers = [None for i in range(num)]
+
+    def put_task(self, data):
```

Comment:
Can we explode the args to `put_task`?  I'm not a huge fan of passing it as one arg, tho then explode it again.  individual args seems a better idea.